### PR TITLE
Refactor and add unit tests for scheduled sc execution

### PR DESCRIPTION
**Pull request by SebastianMarian**
* Added unit tests for scheduled txs execution component
* Refactor transactionsV2

### CODE REVIEW

File: process/block/preprocess/transactionsV2.go [hunk 1]
```diff
@@ -427,160 +282,64 @@ func (txs *transactions) createScheduledMiniBlocks(
...
-	mapGasConsumedByMiniBlockInReceiverShard := txs.initGasConsumed()
-
-	gasInfo := gasConsumedInfo{
-		gasConsumedByMiniBlockInReceiverShard: uint64(0),
-		gasConsumedByMiniBlocksInSenderShard:  uint64(0),
-		totalGasConsumedInSelfShard:           uint64(0),
-	}
-
-	senderAddressToSkip := []byte("")
+	csmbi := txs.initCreateScheduledMiniBlocks()
```
Comment (sasurobert) [inline process/block/preprocess/transactionsV2.go:285]: there is little diference between csmbi and cpmbi from line 204.

File: process/block/preprocess/transactionsV2.go [hunk 2]
```diff
@@ -738,3 +488,313 @@ func getAdditionalTimeFunc() func() bool {
...
+
+	miniBlock.TxHashes = append(miniBlock.TxHashes, txHash)
+	txs.blockSizeComputation.AddNumTxs(1)
+	if scheduledTxMbInfo.isCrossShardScCallTx {
+		if !csmbi.firstCrossShardScCallTxFound {
+			csmbi.firstCrossShardScCallTxFound = true
+			txs.blockSizeComputation.AddNumMiniBlocks(1)
+		}
+		csmbi.mapCrossShardScCallTxs[receiverShardID]++
+		if csmbi.mapCrossShardScCallTxs[receiverShardID] > csmbi.maxCrossShardScCallTxsPerShard {
```
Comment (sasurobert) [inline process/block/preprocess/transactionsV2.go:789]: csmbi.mapCrossShardScCallTxs[receiverShardID] take out into a variable

File: process/block/preprocess/transactionsV2.go [hunk 3]
```diff
@@ -738,3 +488,313 @@ func getAdditionalTimeFunc() func() bool {
...
+	}
+
+	miniBlock.TxHashes = append(miniBlock.TxHashes, txHash)
+	txs.blockSizeComputation.AddNumTxs(1)
+	if txMbInfo.isCrossShardScCallOrSpecialTx {
+		if !cpmbi.firstCrossShardScCallOrSpecialTxFound {
+			cpmbi.firstCrossShardScCallOrSpecialTxFound = true
+			txs.blockSizeComputation.AddNumMiniBlocks(1)
+		}
+		cpmbi.mapCrossShardScCallsOrSpecialTxs[receiverShardID]++
```
Comment (sasurobert) [inline process/block/preprocess/transactionsV2.go:633]: cpmbi.mapCrossShardScCallsOrSpecialTxs[receiverShardID] take out into a separate variable

File: process/block/preprocess/transactionsV2.go [hunk 4]
```diff
@@ -738,3 +488,313 @@ func getAdditionalTimeFunc() func() bool {
...
+		if !ok {
+			log.Error("applyExecutedTransaction.SubBalanceFromAddress",
+				"sender address", tx.GetSndAddr(),
+				"tx max total cost", txMaxTotalCost,
+				"err", process.ErrInsufficientFunds)
+		}
+	}
+
+	isMiniBlockEmpty := len(miniBlock.TxHashes) == 0
+	firstMiniBlockSplitForReceiverShardFound := (txMbInfo.isReceiverSmartContractAddress && cpmbi.mapScsForShard[receiverShardID] == 0 && cpmbi.mapTxsForShard[receiverShardID] > 0) ||
```
Comment (sasurobert) [inline process/block/preprocess/transactionsV2.go:620]: this is an incredibly complicated condition. can we make it easier ?

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: process/block/preprocess/transactionsV2.go [hunk 5]
```diff
@@ -52,213 +20,81 @@ func (txs *transactions) createAndProcessMiniBlocksFromMeV2(
...
-		if err != nil && !errors.Is(err, process.ErrFailedTransaction) {
-			if errors.Is(err, process.ErrHigherNonceInTransaction) {
-				senderAddressToSkip = tx.GetSndAddr()
-			}
-			continue
-		}
-
-		senderAddressToSkip = []byte("")
-
-		gasRefunded := txs.gasHandler.GasRefunded(txHash)
```
Comment (SebastianMarian) [inline process/block/preprocess/transactionsV2.go:179]: This gasRefunded approach should be modified exactly like in PR: https://github.com/ElrondNetwork/elrond-go/pull/3016

File: process/block/preprocess/transactionsV2.go [hunk 6]
```diff
@@ -52,213 +20,81 @@ func (txs *transactions) createAndProcessMiniBlocksFromMeV2(
...
 	log.Debug("createAndProcessMiniBlocksFromMeV2 has been started")
 
-	mapSCTxs := make(map[string]struct{})
-	mapTxsForShard := make(map[uint32]int)
-	mapScsForShard := make(map[uint32]int)
-	mapCrossShardScCallsOrSpecialTxs := make(map[uint32]int)
-	maxCrossShardScCallsOrSpecialTxsPerShard := 0
-
-	processingInfo := processedTxsInfo{}
+	cpmbi := txs.initCreateAndProcessMiniBlocks()
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/block/preprocess/transactionsV2.go:23]: done

File: process/block/preprocess/transactionsV2.go (hunk 2 above)
Comment (SebastianMarian) [reply process/block/preprocess/transactionsV2.go:789]: done

File: process/block/preprocess/transactionsV2.go (hunk 3 above)
Comment (SebastianMarian) [reply process/block/preprocess/transactionsV2.go:633]: done

File: process/block/preprocess/transactionsV2.go (hunk 4 above)
Comment (SebastianMarian) [reply process/block/preprocess/transactionsV2.go:620]: refactored

**Review by AdoAdoAdo — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**